### PR TITLE
magento/magento2#31779:Cart Price rule not working as expected for joined mutation for configurable options update - changed key for validator

### LINKED ISSUE #31779: Cart Price rule not working as expected for joined mutation for configurable options update
**While using joined mutations in graphql for updating configurable items, Cart Price Rules not working as expected**   

### Preconditions (*)

1. latest magento 

### Steps to reproduce (*)
make sure you have active cart price rule (for example 20% OFF Ever $200-plus purchase!*  with magento sampledata )
```
mutation UpdateConfigurableOptions(
  $cartId: String!
  $cartItemId: Int!
  $parentSku: String!
  $variantSku: String!
  $quantity: Float!
) {
  addConfigurableProductsToCart(
    input: {
      cart_id: $cartId
      cart_items: [
        {
          data: { quantity: $quantity, sku: $variantSku }
          parent_sku: $parentSku
        }
      ]
    }
  ) {
    cart {
      id
      total_quantity
      available_payment_methods {
        code
        title
      }
      prices {
        grand_total {
          value
          currency
        }
      }
    }
  }
  removeItemFromCart(input: { cart_id: $cartId, cart_item_id: $cartItemId }) {
    cart {
      id
      total_quantity
      available_payment_methods {
        code
        title
      }
      prices {
        grand_total {
          value
          currency
        }
      }
    }
  }
}

variables
{
  "cartId":"HrDsuEYzcAfHDtP7Cy09Vhf0gbI0VgVU",
  "cartItemId":13,
  "parentSku":"WJ01",
  "variantSku":"WJ01-S-Red",
  "quantity":2
}
```

### Expected result (*)
```
{
  "data": {
    "addConfigurableProductsToCart": {
      "cart": {
        "id": "Pz3Jd1hiXGFXfuKYTGE6hGyDGYJFYAm3",
        "total_quantity": 3,
        "available_payment_methods": [
          {
            "code": "checkmo",
            "title": "Check / Money order"
          }
        ],
        "prices": {
          "grand_total": {
            "value": 180,
            "currency": "USD"
          }
        }
      }
    },
    "removeItemFromCart": {
      "cart": {
        "id": "Pz3Jd1hiXGFXfuKYTGE6hGyDGYJFYAm3",
        "total_quantity": 2,
        "available_payment_methods": [
          {
            "code": "checkmo",
            "title": "Check / Money order"
          }
        ],
        "prices": {
          "grand_total": {
            "value": 150,
            "currency": "USD"
          }
        }
      }
    }
  }
}
```

### Actual result (*)
```
{
  "data": {
    "addConfigurableProductsToCart": {
      "cart": {
        "id": "Pz3Jd1hiXGFXfuKYTGE6hGyDGYJFYAm3",
        "total_quantity": 3,
        "available_payment_methods": [
          {
            "code": "checkmo",
            "title": "Check / Money order"
          }
        ],
        "prices": {
          "grand_total": {
            "value": 180,
            "currency": "USD"
          }
        }
      }
    },
    "removeItemFromCart": {
      "cart": {
        "id": "Pz3Jd1hiXGFXfuKYTGE6hGyDGYJFYAm3",
        "total_quantity": 2,
        "available_payment_methods": [
          {
            "code": "checkmo",
            "title": "Check / Money order"
          }
        ],
        "prices": {
          "grand_total": {
            "value": 120,
            "currency": "USD"
          }
        }
      }
    }
  }
}
```
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by sasha19957099**
### Description (*)
Changed key in sales rules validator to avoid incorrect cached results

### Fixed Issues (if relevant)
1. Fixes https://github.com/magento/magento2/issues/31779

### Manual testing scenarios (*)
1. use the latest magento with the sample data
2. add to cart 1 product (WJ01-S-Yellow)
3. make sure you have active cart price rule (for example 20% OFF Ever $200-plus purchase!* with magento sampledata )
4. Run query
`mutation UpdateConfigurableOptions(`
`  $cartId: String!`
`  $cartItemId: Int!`
`  $parentSku: String!`
`  $variantSku: String!`
`  $quantity: Float!`
`) {`
`  addConfigurableProductsToCart(`
`    input: {`
`      cart_id: $cartId`
`      cart_items: [`
`        {`
`          data: { quantity: $quantity, sku: $variantSku }`
`          parent_sku: $parentSku`
`        }`
`      ]`
`    }`
`  ) {`
`    cart {`
`      id`
`      total_quantity`
`      available_payment_methods {`
`        code`
`        title`
`      }`
`      prices {`
`        grand_total {`
`          value`
`          currency`
`        }`
`      }`
`    }`
`  }`
`  removeItemFromCart(input: { cart_id: $cartId, cart_item_id: $cartItemId }) {`
`    cart {`
`      id`
`      total_quantity`
`      available_payment_methods {`
`        code`
`        title`
`      }`
`      prices {`
`        grand_total {`
`          value`
`          currency`
`        }`
`      }`
`    }`
`  }`
`}`

`variables`
`{`
`  "cartId":"HrDsuEYzcAfHDtP7Cy09Vhf0gbI0VgVU",`
`  "cartItemId":13,`
`  "parentSku":"WJ01",`
`  "variantSku":"WJ01-S-Red",`
`  "quantity":2`
`}`

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by sasha19957099** [comment]
@magento run all tests

**Comment by sasha19957099** [comment]
@magento run all tests

**Comment by sasha19957099** [comment]
@magento run all tests

File: app/code/Magento/SalesRule/Model/Validator.php [hunk 1]
```diff
@@ -154,15 +157,16 @@ public function __construct(
...
      *
-     * @param int $websiteId
-     * @param int $customerGroupId
-     * @param string $couponCode
+     * @param $websiteId
+     * @param CartInterface $quote
      * @return $this
      */
-    public function init($websiteId, $customerGroupId, $couponCode)
+    public function init($websiteId, CartInterface $quote): self
```
Comment (cpartica) [inline app/code/Magento/SalesRule/Model/Validator.php:164]: I'm afraid we can't do this in a patch release. no protected and especially public methods modification. Won't be able to approve SVC
At most we could add an optional parameter which is CartInterface $quote = null
Also don't change the return type, leave self out

**Review by cpartica — CHANGES_REQUESTED** [review]

**Comment by engcom-Charlie** [comment]
Hi @sasha19957099,

Thank you for your contribution!

In reproduction steps of this issue, you are using addConfigurableProductsToCart mutation.

As per [dev docs](https://devdocs.magento.com/guides/v2.4/graphql/mutations/add-configurable-products.html) it is recommended of using [addProductsToCart mutation](https://devdocs.magento.com/guides/v2.4/graphql/mutations/add-products-to-cart.html) rather than addConfigurableProductsToCart mutation.

Hence before proceeding with this PR, in discussion with the PO for the same. Till that time moving this PR to On Hold.

Thank you!

CC: @cpartica

**Comment by engcom-Charlie** [comment]
As per @cpartica confirmation we need to process it ahead so moving this PR back to Changes Requested.

Thank you!

**Comment by engcom-Charlie** [comment]
Hi @sasha19957099,

Thank you for your contribution!

As per the review comments, we can not go with the given solution in this PR as its making modifications in public methods.

image

Do you want to propose any other solution for the linked issue.

Thank you!

**Comment by engcom-Charlie** [comment]
As we didn't got any reply on above [comment](https://github.com/magento/magento2/pull/31838#issuecomment-1096447219), due to inactivity, closing this PR now.

Please reopen and update if you wish to continue.

Thank you for your contribution!

**Pull request closed without merging**